host/xtest/pkcs11_1000.c

```diff
@@ -8016,3 +8133,111 @@ static void xtest_pkcs11_test_1025(ADBG_Case_t *c)
 
 ADBG_CASE_DEFINE(pkcs11, 1025, xtest_pkcs11_test_1025,
 		 "PKCS11: EDDSA key generation and signing");
+
+static CK_RV test_login_logout_acl_auth(ADBG_Case_t *c, CK_SLOT_ID slot)
+{
+	CK_FLAGS session_flags = CKF_SERIAL_SESSION | CKF_RW_SESSION;
+	CK_SESSION_HANDLE session = CK_INVALID_HANDLE;
+	CK_RV rv = CKR_GENERAL_ERROR;
+
+	Do_ADBG_BeginSubCase(c, "Test C_Login()/C_Logout() with ACL based authentication");
+
+	rv = C_OpenSession(slot, session_flags, NULL, 0, &session);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	/* Logout: should fail as we did not log in yet */
+	rv = C_Logout(session);
+	ADBG_EXPECT_CK_RESULT(c, CKR_USER_NOT_LOGGED_IN, rv);
+
+	/* Login/re-log/logout user */
+	rv = C_Login(session, CKU_USER, NULL, 0);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	rv = C_Login(session, CKU_USER, NULL, 0);
+	ADBG_EXPECT_CK_RESULT(c, CKR_USER_ALREADY_LOGGED_IN, rv);
+
+	rv = C_Logout(session);
+	ADBG_EXPECT_CK_OK(c, rv);
+
+out:
+	Do_ADBG_EndSubCase(c, "Test C_Login()/C_Logout() with ACL based authentication");
+	return rv;
+}
+
+#define GID_STR_LEN  13
+
+static void xtest_pkcs11_test_1026(ADBG_Case_t *c)
+{
+	CK_RV rv = CKR_GENERAL_ERROR;
+	CK_FUNCTION_LIST_PTR ckfunc_list = NULL;
+	CK_SLOT_ID slot = 0;
+	CK_TOKEN_INFO token_info = { };
+	char gid_str[GID_STR_LEN] = { };
+
+	rv = C_GetFunctionList(&ckfunc_list);
+	if (!ADBG_EXPECT_CK_OK(c, rv) ||
+	    !ADBG_EXPECT_NOT_NULL(c, ckfunc_list->C_InitToken) ||
+	    !ADBG_EXPECT_NOT_NULL(c, ckfunc_list->C_InitPIN) ||
+	    !ADBG_EXPECT_NOT_NULL(c, ckfunc_list->C_Login) ||
+	    !ADBG_EXPECT_NOT_NULL(c, ckfunc_list->C_Logout))
+		goto out;
+
+	if (setenv("CKTEEC_LOGIN_TYPE", "user", 1))
+		goto out;
+
+	rv = init_lib_and_find_token_slot(&slot, ACL_AUTH);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	rv = C_GetTokenInfo(slot, &token_info);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	/* Abort test if token is about to lock */
+	if (!ADBG_EXPECT_TRUE(c, !(token_info.flags & CKF_SO_PIN_FINAL_TRY)))
+		goto out;
+
+	if (!(token_info.flags & CKF_TOKEN_INITIALIZED)) {
+		rv = test_uninitialized_token(c, slot, ACL_AUTH);
+		if (rv != CKR_OK)
+			goto out;
+	}
+
+	/*
+	 * The current connection is closed and a new CKU_USER connection will be opened.
+	 */
+	rv = close_lib();
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	if (setenv("CKTEEC_LOGIN_TYPE", "group", 1))
```

Comment:
using ADBG_EXPECT_xxx() make nice error trace pointing at the file/line here that areport a failure.
I think this should also check `getenv()`/`setenv()` return vlaue using an `ADBG_EXPECT_xxx()` macro.